runtime/src/snapshot_utils.rs

```diff
@@ -814,23 +1387,66 @@ where
         .next()
 }
 
-pub fn purge_old_snapshot_archives<P: AsRef<Path>>(
+/// Get the path for the incremental snapshot archive with the highest slot, for a given full
+/// snapshot slot, in a directory
+pub fn get_highest_incremental_snapshot_archive_info<P>(
     snapshot_output_dir: P,
-    maximum_snapshots_to_retain: usize,
-) {
+    full_snapshot_slot: Slot,
+) -> Option<IncrementalSnapshotArchiveInfo>
+where
+    P: AsRef<Path>,
+{
+    // Do not call get_sorted_incremental_snapshot_archives() here!  Since we want to filter down
+    // to only the incremental snapshot archives that have the same full snapshot slot as the value
+    // passed in, perform the filtering before sorting to avoid doing unnecessary work.
+    let mut incremental_snapshot_archives = get_incremental_snapshot_archives(snapshot_output_dir)
+        .into_iter()
+        .filter(|incremental_snapshot_archive_info| {
+            incremental_snapshot_archive_info.base_slot == full_snapshot_slot
+        })
+        .collect::<Vec<_>>();
+    sort_incremental_snapshot_archives(&mut incremental_snapshot_archives);
+    incremental_snapshot_archives.into_iter().next()
+}
+
+pub fn purge_old_snapshot_archives<P>(snapshot_output_dir: P, maximum_snapshots_to_retain: usize)
+where
+    P: AsRef<Path>,
+{
     info!(
-        "Purging old snapshots in {:?}, retaining {}",
-        snapshot_output_dir.as_ref(),
+        "Purging old snapshots in {}, retaining {}",
+        snapshot_output_dir.as_ref().display(),
         maximum_snapshots_to_retain
     );
-    let mut archives = get_sorted_snapshot_archives(snapshot_output_dir);
+    let mut archives = get_sorted_snapshot_archives(snapshot_output_dir.as_ref());
     // Keep the oldest snapshot so we can always play the ledger from it.
     archives.pop();
     let max_snaps = max(1, maximum_snapshots_to_retain);
     for old_archive in archives.into_iter().skip(max_snaps) {
+        trace!(
+            "Purging old snapshot archive: {}",
+            old_archive.path.display()
+        );
         fs::remove_file(old_archive.path)
-            .unwrap_or_else(|err| info!("Failed to remove old snapshot: {:}", err));
+            .unwrap_or_else(|err| info!("Failed to remove old snapshot: {}", err));
     }
+
+    // Only keep incremental snapshots for the latest full snapshot
+    // bprumo TODO: As an option to further reduce the number of incremental snapshots, only a
+    // subset of the incremental snapshots for the lastest full snapshot could be kept.  This could
+    // reuse maximum_snapshots_to_retain, or use a new field just for incremental snapshots.
```

Comment:
I plan to address this TODO in a subsequent PR when adding integration tests for incremental snapshots.

Reply:
I think at interface level maximum_snapshots_to_retain is good enough. Snapshot is a snapshot. Incremental or full is implementation detail. for efficiency. The user may be advised that with incremental feature they can afford to keep more snapshots around for the same storage overhead.

Reply:
I've created a new issue to track this TODO, #18639. I think this may need some additional configuration options. Stephen has mentioned wanting to keep old (full) snapshots around longer for replay purposes, and I could see similar configurations requested for incremental snapshots.